### ISSUE TITLE
how to use PMD 5.8.1 Java API to create a rule and to raise a ruleviolation

### ISSUE THREAD

**Issue by Neha-Dhonde** [pmd/pmd#617]
Hi,

I am new to custom rule development using PMD java API. I am trying to create a simple rule which scans for classes for empty package name. From a high level perspective , I believe the code would look something like below (please correct me if I am incorrect).

```java
public class PackageEmptyRule extends AbstractJavaRule  {
	   
           String packageName;

	   public Object visit(ASTPackageDeclaration node, Object data) {
                  packageName = node.getPackageNameImage();
                  return data;
	   }

	   // I chose to listen for class declarations because we are 
	   // guaranteed to have it, and it will always be called after 
	   // the package declaration.
	   public Object visit(ASTClassOrInterfaceType node, Object data) {
	      if (packageName == null) {
	         // Messages printed to standard output will be shown when
	         // PMD is run from the command line or as an Ant target.
	         System.out.println("Found a class with no package declaration.");

                 //Add code here to add a rule violation.
	      }
	      return data;
	   }
}
```

Request you to provide an example of writing a custom rule using java api.

**Comment by jsotuyod** [comment]
@Neha-Dhonde your first approach at the problem seems quite on track.

You are simply missing the `addViolation` call to report the issue back to PMD. You can find [some extra doc on our site](https://pmd.github.io/pmd/pmd_devdocs_writing_pmd_rules.html#write-code-to-add-rule-violations-where-appropriate)

There are a couple of improvements to be done, but mostly nitpicking (such as using the rule chain for performance, which is unfortunately under documented #597 ).

In this case however, I'd personally not go with a Java rule and prefer XPath, mostly due to simplicity.

**Comment by oowekyala** [comment]
@Neha-Dhonde Bear in mind that a single instance of the rule will analyse several files, so you might want to override the `visit(ASTCompilationUnit, Object)` to reset your `packageName` variable to null, otherwise your rule will ignore some violations cases. That method is invoked upon entry to each file (as ASTCompilationUnit is the root of any Java AST), unlike `visit(ASTPackageDeclaration,Object)`, which is only called on PackageDeclaration nodes (and you want to catch missing package declarations)

@jsotuyod is right though, an XPath rule would be simpler. There's a guide [here](https://pmd.github.io/pmd/pmd_devdocs_writing_xpath_rules.html)

**Comment by jsotuyod** [comment]
@oowekyala that's a good catch! I'd personally not visit `ASTCompilationUnit` and prefer using the  `start` method of the rule.

**Comment by oowekyala** [comment]
@jsotuyod TIL that existed, thanks!

**Comment by Neha-Dhonde** [comment]
Thanks @jsotuyod & @oowekyala for your inputs. The reason i am going for a java based custom rule is for scenarios like below:
         I want to scan the files for the usage of Log4j APIs (info(), debug() etc.) being used in the code. So for example a developer creates a class instance of Logger class named as "myLogger". And in one of the methods makes a call as myLogger.info(..). 
        So my java rule should identify that "info()" API of class org.apache.log4j.Logger has been invoked and should be a candidate for violation. However on other hand , if ".info()" method of some custom Util class of the application is called it should be ignored by my java rule.
So the rule check I am trying to work out should represent behavior something like -
         if (instanceOf  org.apache.log4j.Logger && info())

**Comment by oowekyala** [comment]
@Neha-Dhonde That looks like an entirely different rule. It's advisable to create one rule for each type of violation you want to catch, mixing the detection logic for missing package statements and logger calls seems to ask for trouble imho. 

I'd recommend making an XPath rule for your first rule, and another Java rule for logger call detection. You could use the symbol table to find the declaration of the logger and its type, like in [ForLoopCanBeForeach](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java#L82), because unfortunately type resolution still loses some information in calls such as `myLogger.info(..)` (see #497).

**Comment by jsotuyod** [comment]
I believe the original question has already been answered. Shall there be any follow up questions, just open a new issue.
